src/beast/beast/crypto/impl/ripemd_context.h

```diff
@@ -0,0 +1,441 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef BEAST_CRYPTO_RIPEMD_CONTEXT_H_INCLUDED
+#define BEAST_CRYPTO_RIPEMD_CONTEXT_H_INCLUDED
+
+#include <beast/utility/noexcept.h>
+#include <array>
+#include <cstdint>
+#include <cstring>
+
+namespace beast {
+namespace detail {
+
+// Based on
+// https://code.google.com/p/blockchain/source/browse/trunk/RIPEMD160.cpp
+/*
+    Copyright (c) Katholieke Universiteit Leuven
+    1996 All Rights Reserved
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"), to deal
+    in the Software without restriction, including without limitation the rights
+    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    THE SOFTWARE.
+*/
+// This implementation has been modified from the
+// original. It has been updated for C++11.
+
+struct ripemd160_context
+{
+    static unsigned int const block_size = 64;
+    static unsigned int const digest_size = 20;
+
+    unsigned int tot_len;
+    unsigned int len;
+    std::uint32_t block[64];
+    std::uint32_t h[5];
+};
+
+// ROL(x, n) cyclically rotates x over n bits to the left
+// x must be of an unsigned 32 bits type and 0 <= n < 32.
+#define BEAST_RIPEMD_ROL(x, n)  (((x) << (n)) | ((x) >> (32-(n))))
+
+// the five basic functions F(), G() and H()
+#define BEAST_RIPEMD_F(x, y, z)  ((x) ^  (y) ^ (z)) 
+#define BEAST_RIPEMD_G(x, y, z) (((x) &  (y)) | (~(x) & (z))) 
+#define BEAST_RIPEMD_H(x, y, z) (((x) | ~(y)) ^ (z))
+#define BEAST_RIPEMD_I(x, y, z) (((x) &  (z)) | ((y) & ~(z))) 
+#define BEAST_RIPEMD_J(x, y, z)  ((x) ^  ((y) | ~(z)))
+  
+// the ten basic operations FF() through III()
+#define BEAST_RIPEMD_FF(a, b, c, d, e, x, s) {      \
+      (a) += BEAST_RIPEMD_F((b), (c), (d)) + (x);   \
+      (a) = BEAST_RIPEMD_ROL((a), (s)) + (e);       \
+      (c) = BEAST_RIPEMD_ROL((c), 10); }
+#define BEAST_RIPEMD_GG(a, b, c, d, e, x, s) {      \
```

Comment:
All this is a complete mystery to me.  Where do these equations come from?  Where do these magical numbers come from?  Could you link to some sort of spec?